src/database/database-files.cpp

```diff
@@ -177,3 +178,113 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
 		res.emplace_back(player.getName());
 	}
 }
+
+AuthDatabaseFiles::AuthDatabaseFiles(const std::string &savedir) : m_savedir(savedir)
+{
+	readAuthFile();
+}
+
+bool AuthDatabaseFiles::getAuth(const std::string &name, AuthEntry &res)
+{
+	const auto res_i = m_auth_list.find(name);
+	if (res_i == m_auth_list.end()) {
+		return false;
+	}
+	res = res_i->second;
+	return true;
+}
+
+bool AuthDatabaseFiles::saveAuth(const AuthEntry &authEntry)
+{
+	m_auth_list[authEntry.name] = authEntry;
+
+	// save entire file
+	return writeAuthFile();
+}
+
+bool AuthDatabaseFiles::createAuth(AuthEntry &authEntry)
+{
+	m_auth_list[authEntry.name] = authEntry;
+
+	// save entire file
+	return writeAuthFile();
+}
+
+bool AuthDatabaseFiles::deleteAuth(const std::string &name)
+{
+	auto count = m_auth_list.erase(name);
+	if (!count) {
+		// did not delete anything -> hadn't existed
+		return false;
+	}
+	return writeAuthFile();
+}
+
+void AuthDatabaseFiles::listNames(std::vector<std::string> &res)
+{
+	res.clear();
+	res.reserve(m_auth_list.size());
+	for (auto const &res_pair : m_auth_list) {
+		res.push_back(res_pair.first);
+	}
+	return;
+}
+
+void AuthDatabaseFiles::reload()
+{
+	readAuthFile();
+	return;
+}
+
+bool AuthDatabaseFiles::readAuthFile()
+{
+	std::string path = m_savedir + DIR_DELIM + "auth.txt";
+	std::ifstream file(path, std::ios::binary);
+	if (!file.good()) {
+		return false;
+	}
+	m_auth_list.clear();
+	while (file.good()) {
+		std::string line;
+		std::getline(file, line);
+		std::vector<std::string> parts = str_split(line, ':');
+		if (parts.size() < 4) // also: empty line at end
```

Comment:
Just to have it said: Last login exists since almost four years but our auth mechanism didn't add dummy values if they didn't exist.
However, I think we can consider this as a feature to clean up really old, unused accounts.

Reply:
I don't understand your comment, sorry. Do you mean that there are `auth.txt` files without the fourth field? I think the 4 was just meant as some kind of sanity check to guard against empty lines. Would it help if I changed it to 3? But what would I use as `last_login` then?

Reply:
IMO it's not okay for Minetest to quietly drop data, even if it's okay and even useful in most cases

Reply:
These players are however still in the player database, so when their auth is deleted, someone could take over this used account. Result: Joining at any unknown place with lack of initial items or full-equipped armor.
@bendeutsch Could you please change this to `parts.size() < 3` and set `last_login` to `0` when it's not available?

Reply:
yes it's a good idea @SmallJoker

Reply:
Done. Sorry for the delay :-(